packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,413 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin, NoopLogger, isValid } from '@opentelemetry/core';
+import {
+  Span,
+  SpanKind,
+  SpanOptions,
+  Logger,
+  Tracer,
+  Attributes,
+} from '@opentelemetry/types';
+import {
+  ClientRequest,
+  IncomingMessage,
+  request,
+  RequestOptions,
+  ServerResponse,
+} from 'http';
+import * as semver from 'semver';
+import * as shimmer from 'shimmer';
+import * as url from 'url';
+import {
+  HttpPluginConfig,
+  Http,
+  Func,
+  ResponseEndArgs,
+  ParsedRequestOptions,
+  HttpRequestArgs,
+} from './types';
+import { Format } from './enums/format';
+import { AttributeNames } from './enums/attributeNames';
+import { Utils } from './utils';
+
+/**
+ * Http instrumentation plugin for Opentelemetry
+ */
+export class HttpPlugin extends BasePlugin<Http> {
+  static readonly component = 'http';
+  options!: HttpPluginConfig;
+  protected _logger!: Logger;
+  protected readonly _tracer!: Tracer;
+
+  constructor(public moduleName: string, public version: string) {
+    super();
+    // TODO: remove this once a logger will be passed
+    // https://github.com/open-telemetry/opentelemetry-js/issues/193
+    this._logger = new NoopLogger();
+  }
+
+  /** Patches HTTP incoming and outcoming request functions. */
+  protected patch() {
+    this._logger.debug(
+      'applying patch to %s@%s',
+      this.moduleName,
+      this.version
+    );
+
+    shimmer.wrap(
+      this._moduleExports,
+      'request',
+      this._getPatchOutgoingRequestFunction()
+    );
+
+    // In Node 8-10, http.get calls a private request method, therefore we patch it
+    // here too.
+    if (semver.satisfies(this.version, '>=8.0.0')) {
+      shimmer.wrap(
+        this._moduleExports,
+        'get',
+        this._getPatchOutgoingGetFunction()
+      );
+    }
+
+    if (
+      this._moduleExports &&
+      this._moduleExports.Server &&
+      this._moduleExports.Server.prototype
+    ) {
+      shimmer.wrap(
```

Comment:
If I understand this correct this adds an `emit` property on `Server.prototype` and the implementation of our `emit` function calls the original emit (from `EventEmitter`) as seen on path time.
If there is some other modules which monkey patches `emit` on EventEmitter afterwards this monkey patched version is skipped as it was not seen at patch time.

Reply:
I hacked a sample to illustrate this. The sequence of calling `patchHttp` and `patchEmitter()` matters but in an ideal world it should not.

```
"use strict"

const http = require("http");
const shimmer = require("shimmer");

function patchEmitter() {
	const { EventEmitter } = require("events");

	const orgEmit = EventEmitter.prototype.emit;
	EventEmitter.prototype.emit = function theEmit(e) {
		if (e === "request") console.log(`theEmit ${e}`);
		return orgEmit.apply(this, arguments);
	}
}

function patchHttp() {
	shimmer.wrap(http.Server.prototype, "emit", getHttpEmit());

	function getHttpEmit() {
		return (original) => {
			return myHttpEmit(original);
		}
	}

	function myHttpEmit(original) {
		return function theHttpEmit(e) {
			if (e === "request") console.log(`theHttpEmit ${e}`)
			return original.apply(this, arguments);
		}
	}
}

// The sequence here matters!
patchHttp();
patchEmitter(); 

http.createServer((req, res) => {
	console.log("server request");
	res.end();
}).listen(8000);

http.get("http://localhost:8000/");
```

Reply:
I don't agree here, it's not about patching too late in the sense after a server starts or similar it's the sequence between OT and others.
This is in special problematic as we tell users to require OT first so if they follow the guideline some other modules doing patching may be broken.

By adding `emit` in `Server.prototype` the inheritance structure is changed and this is a visible side effect monitoring should avoid and this does not happen with the "usual" monkey patching which just replaces an existing function by another.
In the usual replace function patching way the sequence is don't care - either the one calls the other or the other way around but always both are called. But here we omit one.

I'm not sure about the best solution here, one would be an `instanceof Server` check, another one would be to monitor registration of 'request' listeners and a third one would be register our own 'request' listener.

Reply:
created https://github.com/open-telemetry/opentelemetry-js/issues/224
Sure, non of my comments should block the PR as all of them can be fixed later - or ignored.